editors/code/language-configuration.json

```diff
@@ -0,0 +1,39 @@
+{
+    "comments": {
+        "lineComment": "//",
+        "blockComment": [ "/*", "*/" ]
+    },
+    "brackets": [
+        ["{", "}"],
+        ["[", "]"],
+        ["(", ")"],
+        ["<", ">"]
+    ],
+    "autoClosingPairs": [
+        { "open": "{", "close": "}" },
+        { "open": "[", "close": "]" },
+        { "open": "(", "close": ")" },
+        { "open": "<", "close": ">" },
```

Comment:
Will this work correctly (not adding `>`) when typing less then? `if idx < xs.len()` and such?

Reply:
It won't. Unfortunately the only trigger available in the language config API is the character to the right of the current cursor position (`autoCloseBefore`, [docs here](https://code.visualstudio.com/api/language-extensions/language-configuration-guide#autoclosing-before)). If there were an equivalent `autoCloseAfter`, I think we could rely on whitespace to the left of the cursor position as an indicator that auto-close should be disabled, but that option doesn't exist. An issue was opened with VS Code to request `autoCloseAfter` functionality (specifically because of Rust):  https://github.com/microsoft/vscode/issues/109838, but the VS Code team is not receptive to adding it. The rationale behind `autoCloseBefore` is that placing a bracket in the middle of a group of characters likely means you are going to navigate somewhere else and close it (the manual equivalent of `autoSurround`). It isn't really targeted at symbols that could have 2 completely different meanings. 

So the question is whether Rust uses angle brackets enough that the occasional extra delete step (for `<` as LT) is a reasonable tradeoff for all the `>` you don't have to type. TypeScript doesn't auto-close angle brackets, despite being VS Code's flagship language, but the volume of angle bracket use is definitely less than Rust.

The OP on #6451 mentioned that RLS already  has auto-closing of angle brackets, but there has not been much consensus about whether that is a good thing. There is currently an open PR to revert the change: https://github.com/rust-lang/vscode-rust/pull/875, which links to several issues where it is discussed (mostly by those who find auto-close obtrusive).

The intermediate option is to allow auto-surround, but not auto-close. 

I'm not sure how much effort is involved to do this correctly in the language server, which is obviously a consideration.

Reply:
We should have auto-surround for `<`, we shouldn't have auto close.

* somtimes you need to add `>`, and
* sometimes you need to remove `>`

are not symmetric -- the latter has a way higher cost than the former. 

Ideally, we should move this all into the server of course, but that's not possible via built-in LSP functionality, *and* (what's worse) there's no VS Code API to do that, and I don't think a hacky solution like we did for `Enter` is worth the hassle at this stage.